### PR TITLE
Geoloc2

### PR DESCRIPTION
**Pull request by tine-krueger**
**FE**
Geolocating via API in filter form with autocomplete for location suggestions.
Area Search
general styling updates

**BE** 
changed filter logic for location and area search

### CODE REVIEW

File: moskito-backend/src/Service/ClosestCampsitesService.php [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    const EARTH_RADIUS = 6371;
+
+    private array $closestCampsites = [];
+    
+    public function getClosestCampsites( array $filter, array $sortedCampsites): array {
+
+        $latFrom = deg2rad($filter['latitude']);
+        $lonFrom = deg2rad($filter['longitude']);
+
+        foreach ($sortedCampsites as $campsite) {
```
Comment (web-devil) [inline moskito-backend/src/Service/ClosestCampsitesService.php:17]: hier wären tests ja super :)

File: moskito-frontend/src/services/getGeocode.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
+
+export function getGeocode(location) {
+
+    const lowerCaseLocation = location.toLowerCase()
+    const urlLocation = lowerCaseLocation.replace(/\s+/g, "+" )
+    const apiKey = process.env.REACT_APP_HERE_API_KEY
```
Comment (web-devil) [inline moskito-frontend/src/services/getGeocode.js]: _HERE_ streichen

**Review by web-devil — CHANGES_REQUESTED** [review]

File: moskito-backend/src/Service/ClosestCampsitesService.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+class ClosestCampsitesService {
+
+    const EARTH_RADIUS = 6371;
+
+    private array $closestCampsites = [];
+    
+    public function getClosestCampsites( array $filter, array $sortedCampsites): array {
+
+        $latFrom = deg2rad($filter['latitude']);
```
Comment (alex-jonas) [inline moskito-backend/src/Service/ClosestCampsitesService.php:14]: Nicely implemented. However, abbreviations lat/lng should be avoided here (...as I was told).

File: moskito-backend/src/Service/ClosestCampsitesService.php [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        $lonFrom = deg2rad($filter['longitude']);
+
+        foreach ($sortedCampsites as $campsite) {
+            $latTo = deg2rad($campsite->getLatitude());
+            $lonTo = deg2rad($campsite->getLongitude());
+
+            $latDelta = $latDelta = $latTo - $latFrom;
+            $lonDelta = $lonTo - $lonFrom;
+
+            $angle = 2 * asin(sqrt(pow(sin($latDelta / 2), 2) + cos($latFrom) * cos($latTo) * pow(sin($lonDelta / 2), 2)));
```
Comment (alex-jonas) [inline moskito-backend/src/Service/ClosestCampsitesService.php:24]: I like this formula :)

File: moskito-frontend/src/components/Forms/FilterCampsiteForm.js [hunk 5]
```diff
@@ -33,16 +42,52 @@ export default function FilterCampsiteForm({getCampsites}) {
         getCampsites(fields)
         history.push('/campsites')
     }
+
+    function handleClick(suggestion) {
+        const place = suggestion.address.postalCode + " " + suggestion.address.city
+        getGeocode(place)
+        .then(result => {
+            const geocodes = result.Response.View[0].Result[0].Location.DisplayPosition
+            setValues(
```
Comment (alex-jonas) [inline moskito-frontend/src/components/Forms/FilterCampsiteForm.js:53]: The naming (values, fields) appears to be a bit generic.

File: moskito-frontend/src/components/Forms/FilterCampsiteForm.js [hunk 6]
```diff
@@ -33,16 +42,52 @@ export default function FilterCampsiteForm({getCampsites}) {
...
+                {
+                    ...fields,
+                    latitude: geocodes.Latitude,
+                    longitude: geocodes.Longitude,
+                    postalCode: place
+                }
+            )
+        })
+        .catch(error => console.log('error', error));
```
Comment (alex-jonas) [inline moskito-frontend/src/components/Forms/FilterCampsiteForm.js]: The error handling could be improved for the user.

File: moskito-frontend/src/components/Forms/FilterCampsiteForm.js [hunk 7]
```diff
@@ -33,16 +42,52 @@ export default function FilterCampsiteForm({getCampsites}) {
...
+        setValues({
+            ...fields,
+            [event.target.name]: event.target.value
+        })
+        fields.postalCode.length > 2 ?
+        autocomplete(fields.postalCode)
+        .then(results => setSuggestions(results.suggestions)) 
+        .catch(() =>
+          setSuggestions([
+            { description: 'Service nicht verfügbar', googlePlaceId: 'error' },
```
Comment (alex-jonas) [inline moskito-frontend/src/components/Forms/FilterCampsiteForm.js]: What do you need the googlePlaceId for? ;-)

File: moskito-frontend/src/components/Forms/InputField.js [hunk 8]
```diff
@@ -6,12 +6,14 @@ InputField.propTypes = {
...
-    placeholder: PropTypes.string
+    placeholder: PropTypes.node,
+    children: PropTypes.string
 }
 
-export default function InputField({type, name, value, onChange, placeholder}) {
+export default function InputField({type, name, value, onChange, placeholder, children, margin}) {
     return (
-        <Label>
+        <Label margin={margin}>
```
Comment (alex-jonas) [inline moskito-frontend/src/components/Forms/InputField.js]: Maybe this prop could be named more meaningful, less technically?

File: moskito-frontend/src/components/Forms/Suggestions.js [hunk 9]
```diff
@@ -0,0 +1,27 @@
+import styled from 'styled-components/macro'
+
+export default function Suggestions({suggestions, onClick}) {
+    return (
+        <ListStyled>{suggestions.map((suggestion, index) =>  <ListElement key={index} onClick={() => onClick(suggestion)}>{suggestion.label}</ListElement>)}</ListStyled>
```
Comment (alex-jonas) [inline moskito-frontend/src/components/Forms/Suggestions.js]: A line break would be nice here. What about prettier?

File: moskito-frontend/src/services/getInitialCampsiteFilter.js [hunk 10]
```diff
@@ -6,6 +6,9 @@ export default function getInitialCampsiteFilter() {
             name: 'Postleitzahl/Ort',
             dbName: 'postalCode',
             postalCode: '',
+            distance: 100,
+            latitude: -1,
```
Comment (alex-jonas) [inline moskito-frontend/src/services/getInitialCampsiteFilter.js]: Interesting that you start in the south atlantic ocean :-)

File: moskito-frontend/src/components/Navigation/Navigation.js [hunk 11]
```diff
@@ -61,4 +69,17 @@ const NavListItem = styled.li`
...
\ No newline at end of file
+    button svg{
+        width: 2.3em;
+        height: 2.3em;
+    }
+
+    &:hover,
+    &:active,
+    &:focus {
+        background: linear-gradient(145deg, #A2B2B0, #E2FAF6);
```
Comment (alex-jonas) [inline moskito-frontend/src/components/Navigation/Navigation.js:80]: There is a view that color variables should be used where possible. (I know it's hard)

File: moskito-backend/src/Service/ClosestCampsitesService.php (hunk 3 above)
Comment (tine-krueger) [reply moskito-backend/src/Service/ClosestCampsitesService.php:14]: why?

File: moskito-frontend/src/components/Forms/FilterCampsiteForm.js (hunk 7 above)
Comment (tine-krueger) [reply moskito-frontend/src/components/Forms/FilterCampsiteForm.js]: ;) must be a copy/paste error

File: moskito-frontend/src/services/getInitialCampsiteFilter.js (hunk 10 above)
Comment (tine-krueger) [reply moskito-frontend/src/services/getInitialCampsiteFilter.js]: oh good point, didn't know, that geocodes can be negative as well

File: moskito-backend/src/Service/ClosestCampsitesService.php (hunk 3 above)
Comment (alex-jonas) [reply moskito-backend/src/Service/ClosestCampsitesService.php:14]: I got this feedback for my identical code. (But to be honest I didn't change it either.)

Comment (tine-krueger) [reply moskito-backend/src/Service/ClosestCampsitesService.php:14]: will go for that as well...

**Pull request merged**